Java+DB docs [skip ci] (#4486)
openshift/odo: docs/public/deploying-java-app-with-database.adoc

= Binding a Java microservices JPA application to an in-cluster Operator-Managed PostgreSQL Database on minikube

This scenario illustrates binding an `odo` managed Java MicroServices JPA application to an in-cluster Operator managed PostgreSQL Database in the minikube environment.

== Installation and Configuration of the minikube environment
[NOTE]
====
The guide only works with minikube configured with kubernetes 1.19.x or lower. `odo link` cannot link services successfully in a Kubernetes 1.20.x or newer environment.
====

It is recommended that users of this guide obtain a suitable system for running minikube with kubernetes. In practice this should be a 4 core system minimum. Before proceeding to the sample application, please follow the instructions for establishing a minikube environment:

> === Installing and starting docker 

Reviewer: This might need to be removed if we decide to go ahead with my idea of starting minikube as a VM instead of using docker driver.

Author: @dharmit right, I forgot to commit the changes in that Java doc. -_- Apologies, fixed.

To install docker follow link:https://docs.docker.com/engine/install/[the Docker documentation].

=== Installing, starting and configuring minikube

==== Installing minikube
To install minikube follow link:https://minikube.sigs.k8s.io/docs/start/[minikube installation instructions] for your operating system.

==== Starting minikube

* If you are trying to run minikube as a root user, minikube throws an error that docker should not be run as root as a matter of practice and aborts the start up. To proceed, start minikube in a manner which will override this protection mechanism:
+
[source,sh]
----
$ minikube start --force --driver=docker --kubernetes-version=v1.19.8
----

* If you are a non-root user, start minikube as usual:
+
[source,sh]
----
$ minikube start --kubernetes-version=v1.19.8
----

==== Configuring minikube
The application requires an ingress add-on to allow the routes to be created easily. Configure minikube for ingress by adding ingress as a minikube add-on:
[source,sh]
----
> minikube addons enable ingress
----

> It is possible that you face the Dockerhub pull rate limit if you do not have a pull secret for your personal free Dockerhub account. During ingress initialization two of the job pods used by ingress may fail to initialize due to pull rate limits. If this happens, and ingress fails to enable, you add a secret for the pulls for the following service accounts:

Reviewer: Pull rate limit is appearing in both this and the minikube setup doc. We should have it at only place, no?

+
[source,sh]
----
$ kubectl patch serviceaccount ingress-nginx-admission -p '{"imagePullSecrets": [{"name": "regcred"}]}'
$ kubectl patch serviceaccount ingress-nginx -p '{"imagePullSecrets": [{"name": "regcred"}]}'
----

 Default Service Account Pull Secret patch:
 Much like the ingress service accounts, the default service account will need to be patched with a pull secret configured for your personal docker account. 

* switch to the default context:
+
[source,sh]
----
$ kubectl config set-context --current --namespace=default
----

* create the same docker-registry secret configured for your docker , now for the default minikube context:
+
[source,sh]
----
$ kubectl create secret docker-registry regcred --docker-server=<your-registry-server> --docker-username=<your-name> --docker-password=<your-pword> --docker-email=<your-email>
----
+
where:
+
** <your-registry-server> is the DockerHub Registry FQDN. (https://index.docker.io/v1/)
** <your-name> is your Docker username.
** <your-pword> is your Docker password.
** <your-email> is your Docker email.

* Add this new cred ('regcred' in the example above) to the default service account in minikube:
+
[source,sh]
----

> $ kubectl patch serviceaccount default -p '{"imagePullSecrets": [{"name": "regcred"}]}'

Reviewer: Everything up to this point and my previous comment about "pull rate limiting appearing twice" is repeated in the two docs, IIUC.

----

> Kubernetes Dashboard graphical UI config:<br>

Reviewer: This `<br>` shows as-is even in the rich diff view of PR. I hope it won't show up in the docs.

Reviewer: This is still an issue.

> It is helpful to make use of the basic kubernetes dashboard UI to interact with the various kubernetes entities in a graphical way. Refer to the directions [here](https://minikube.sigs.k8s.io/docs/handbook/dashboard/) for enabling and starting the dashboard. Please note, this require the installation of and access to a desktop environment in order to make use of the dashboard. (GNOME + xrdb for example)

Reviewer: TBH, we don't need Kubernetes Dashboard for odo related things. We can skip this.

----

==== Installing the database operator

. Run the following `kubectl` command to make the PostgreSQL Operator available in `my-postgresql-operator-dev4devs-com` namespace of your minikube cluster:
+
[source,sh]
----
$ kubectl create -f https://operatorhub.io/install/postgresql-operator-dev4devs-com.yaml
----

**NOTE**: The `my-postgresql-operator-dev4devs-com` Operator will be installed in the `my-postgresql-operator-dev4devs-com` namespace and will be usable from this namespace only.

==== Creating a database to be used by the sample application
Since the PostgreSQL Operator we installed in above step is available only in `my-postgresql-operator-dev4devs-com` namespace, ensure that `odo` uses this namespace to perform any tasks:
[source,sh]
----
$ odo project set my-postgresql-operator-dev4devs-com
----

You can use the default configurations of the PostgreSQL Operator to start a Postgres database from it. But since our app uses few specific configuration values, lets make sure they are properly populated in the database service we start.

. Store the YAML of the service in a file:
+
[source,sh]
----

> $ odo service create postgresql-operator.v0.1.1/Database --dry-run $ db.yaml

Reviewer: ```suggestion
$ odo service create postgresql-operator.v0.1.1/Database --dry-run > db.yaml
```

----

> . Modify and add following values under `metadata:`section in the `db.yaml` file:

Reviewer: ```suggestion
. Modify and add following values under `metadata:` section in the `db.yaml` file:
```

//: <> (```shell)
//: <> (>  odo service create postgresql-operator.v0.1.1/Database)
//: <> (```)
//: <> (push this service instance to the cluster)
//: <> (```shell)
//: <> (> odo push)
//: <> (```)

. List the service associated with the database created via the PostgreSQL Operator:
+
[source,sh]
----
$ odo service list
NAME                        AGE
Database/sampledatabase     6m31s
----

. Create a Service Binding Request between the application and the database using the Service Binding Operator service created in the previous step `odo link` command: 
+
[source,sh]
----
$ odo link Database/sampledatabase
----

. Push this link to the cluster:
+
[source,sh]
----
$ odo push
----
+
After the link has been created and pushed a secret will have been created containing the database connection data that the application requires.
+
You can inspect the new intermediate secret via the dashboard console in the 'my-postgresql-operator-dev4devs-com' namespace by navigating to Secrets and clicking on the secret named `mysboproj-database-sampledatabase` Notice it contains 4 pieces of data all related to the connection information for your PostgreSQL database instance.
+

> . Push the newly created link. This will terminate the existing application pod and start a new application pod.

Reviewer: Not sure why we need this `odo push`. We already did `odo push` once after doing `odo link`. That should suffice. Another `odo push` is needed only if we're making any change to the application which, as far as I can see, we aren't.

@ajm01 any idea?

Reviewer 2: Hi @dharmit - apologies, I didn't see these updates in my private gmail ... i think that second odo push is extraneous...so not needed at that point in the doc.

Reviewer: @ajm01 OK, thanks. I'll remove that in an upcoming PR. :+1:

+
[source,sh]
----
odo push 
----
+
. Once the new pod has initialized you can see the secret database connection data as it is injected into the pod environment by executing the following:
+
[source,sh]
----
$ odo exec -- bash -c 'export | grep DATABASE'
declare -x DATABASE_CLUSTERIP="10.106.182.173"
declare -x DATABASE_DB_NAME="sampledb"
declare -x DATABASE_DB_PASSWORD="samplepwd"
declare -x DATABASE_DB_USER="sampleuser"
----
+
Once the new version is up (there will be a slight delay until application is available), navigate to the CreatePerson.xhtml using the URL created in a previous step. Enter requested data and click the *Save* button
//![Create Person xhtml page](../../assets/createPersonDB.png)
+
Notice you are re-directed to the `PersonList.xhtml` page, where your data is displayed having been input to the postgreSQL database and retrieved for display purposes.
//![Create Person xhtml page](../../assets/displayPeople.png)
+
You may inspect the database instance itself and query the table to see the data in place by using the postgreSQL command line tool, psql.

> . Navigate to the pod containing your db from the Openshift Console

Reviewer: ```suggestion
. Navigate to the pod containing your db from the Kubernetes Dashboard
```
